### PR TITLE
🚨 Speed up detr image processing

### PR DESCRIPTION
**Pull request by guarin**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48066) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48066)

# What does this PR do?

Speeds up preprocessing on GPU for all detr models by using `group_images_by_shape`. With batch size 32 and ragged inputs between 512px to 2048px I got 27-38% speed up on an H100:

model | baseline step ms | speedup-branch step ms | improvement
-- | -- | -- | --
rtdetrv2 | 20.84 | 15.23 | 27% faster
rf-detr | 25.85 | 16.37 | 37% faster
detr-panoptic | 28.19 | 17.94 | 36% faster
conditional-detr | 28.06 | 17.93 | 36% faster
dab-detr | 28.77 | 17.97 | 38% faster
deformable-detr | 28.16 | 18.01 | 36% faster
lw-detr | 22.68 | 16.26 | 28% faster

Other minor changes:
* `group_images_by_shape` now allows passing `None` as `paired_input` which simplifies annotation handling
* remove rtdetr preprocess and inherit from detr instead (they are identical) 
* 🚨 use antialias=False in rf-detr resize following fix that was recently added in the original rf-detr repo: https://github.com/roboflow/rf-detr/blob/5547a9360c19df7ff8cedcd29f841f7de01f2f75/src/rfdetr/detr.py#L2480-L2481

Looks like a lot of changes but only detr, rt-detr, and rf-detr changed. The rest is auto-generated.

cc @molbap @zucchini-nlp

### CODE REVIEW

**Comment by guarin** [comment]
run-slow: conditional_detr, deformable_detr, detr, grounding_dino, rf_detr, rt_detr, yolos

**Comment by guarin** [comment]
run-slow: rf_detr

File: tests/models/rf_detr/test_modeling_rf_detr.py [hunk 1]
```diff
@@ -461,19 +461,19 @@ def test_inference_object_detection(self):
         label_expectations = Expectations(
             {
                 ("cuda", (8, 0)): [17, 17, 75, 75, 63],
-                ("xpu", None): [17, 17, 75, 75, 63],
+                ("cpu", None): [17, 17, 75, 75, 63],
```
Comment (guarin) [inline tests/models/rf_detr/test_modeling_rf_detr.py:465]: didn't have access to xpu, added cpu instead

File: tests/models/rf_detr/test_modeling_rf_detr.py [hunk 2]
```diff
@@ -461,19 +461,19 @@ def test_inference_object_detection(self):
...
             }
         )
         score_expectations = Expectations(
             {
-                ("cuda", (8, 0)): [0.959521, 0.931229, 0.897797, 0.7286, 0.672863],
-                ("xpu", None): [0.959521, 0.931229, 0.897797, 0.7286, 0.672863],
+                ("cuda", (8, 0)): [0.960207, 0.934029, 0.896868, 0.718928, 0.707891],
+                ("cpu", None): [0.960285, 0.934246, 0.896883, 0.718877, 0.707271],
```
Comment (guarin) [inline tests/models/rf_detr/test_modeling_rf_detr.py:472]: regenerated because of antialiasing change in preprocessing

File: src/transformers/models/conditional_detr/image_processing_conditional_detr.py [hunk 3]
```diff
@@ -656,17 +658,20 @@ def pad(
...
+                    )
+                    for annotation in annotations
+                ]
 
-        # Make a pixel mask for the image, where 1 indicates a valid pixel and 0 indicates padding.
-        pixel_mask = torch.zeros(padded_size, dtype=torch.int64, device=image.device)
-        pixel_mask[: original_size[0], : original_size[1]] = 1
+        # Make a pixel mask for each image, where 1 indicates a valid pixel and 0 indicates padding.
+        pixel_mask = torch.zeros((images.shape[0], *padded_size), dtype=torch.int64, device=images.device)
+        pixel_mask[:, : original_size[0], : original_size[1]] = 1
```
Comment (zucchini-nlp) [inline src/transformers/models/conditional_detr/image_processing_conditional_detr.py]: hmm, adding a leading dim is kinda breaking. 

I dont think anyone would use an `image_proc.pad()` directly, but still can we keep the shape as it was

File: src/transformers/models/conditional_detr/image_processing_conditional_detr.py [hunk 4]
```diff
@@ -735,69 +741,74 @@ def _preprocess(
 
         data = {}
 
-        processed_images = []
-        processed_annotations = []
-        pixel_masks = []  # Initialize pixel_masks here
-        for image, annotation in zip(images, annotations if annotations is not None else [None] * len(images)):
-            # prepare (COCO annotations as a list of Dict -> CONDITIONAL_DETR target as a single Dict per image)
-            if annotations is not None:
-                annotation = self.prepare_annotation(
```
Comment (zucchini-nlp) [inline src/transformers/models/conditional_detr/image_processing_conditional_detr.py:744]: could be an mlinter rule asking users to group images always, unless there is a reason to skip. But prob users will skip anyway when a non-standard model is shipped 😢

File: src/transformers/models/rf_detr/modular_rf_detr.py [hunk 5]
```diff
@@ -128,71 +131,75 @@ def _preprocess(
...
+            processed_images = reorder_images(grouped_images, grouped_images_index)
+            data.update({"pixel_mask": torch.stack(reorder_images(grouped_masks, grouped_images_index), dim=0)})
+
+        data.update({"pixel_values": torch.stack(processed_images, dim=0)})
         encoded_inputs = BatchFeature(data, tensor_type=return_tensors)
         if annotations is not None:
+            annotations = reorder_images(grouped_annotations, grouped_images_index)
             encoded_inputs["labels"] = [
                 BatchFeature(annotation, tensor_type=return_tensors) for annotation in annotations
             ]
```
Comment (zucchini-nlp) [inline src/transformers/models/rf_detr/modular_rf_detr.py:205]: what is the diff here from `DetrImageProcessor._preprocess`?

File: src/transformers/image_transforms.py [hunk 6]
```diff
@@ -1021,14 +1034,15 @@ def group_images_by_shape(
...
-                {key: item.unsqueeze(0) for key, item in _iterate_items(paired_list, is_nested)}
-                for paired_list in paired_inputs
-            ],
-            grouped_images_index,
-        )
+        grouped_images = {key: img.unsqueeze(0) for key, img in _iterate_items(images, is_nested)}
+        paired_grouped_values = [
+            dict.fromkeys(grouped_images, None)
+            if paired_list is None
+            else {key: [item] for key, item in _iterate_items(paired_list, is_nested)}
```
Comment (zucchini-nlp) [inline src/transformers/image_transforms.py:1041]: why we had to return a `list[item]` instead of unsqueezing a new dim?

File: src/transformers/models/rf_detr/image_processing_rf_detr.py [hunk 7]
```diff
@@ -516,71 +522,75 @@ def _preprocess(
...
-                resized_image = self.resize(image, size=size, resample=resample)
-                if annotations is not None:
-                    annotation = self.resize_annotation(
-                        annotation,
-                        orig_size=image.size()[-2:],
-                        target_size=resized_image.size()[-2:],
-                    )
-                image = resized_image
+                orig_size = stacked_images.shape[-2:]
+                stacked_images = self.resize(stacked_images, size=size, resample=resample, antialias=False)
```
Comment (zucchini-nlp) [inline src/transformers/models/rf_detr/image_processing_rf_detr.py:557]: this is breaking, why we need to change `antialias=False`?

File: tests/models/rf_detr/test_modeling_rf_detr.py (hunk 1 above)
Comment (zucchini-nlp) [reply tests/models/rf_detr/test_modeling_rf_detr.py:465]: I dont think we can delete xpu, let's add CPU option and keep XPU

**Review by zucchini-nlp — COMMENTED** [review]
It's great to see grouped iter (could be an mlinter rule 🤔 ). I am seeing a breaking changes, and not sure why are they added, left comments 👇🏻

File: tests/models/rf_detr/test_modeling_rf_detr.py (hunk 1 above)
Comment (guarin) [reply tests/models/rf_detr/test_modeling_rf_detr.py:465]: ~~Ok, even if it fails? Or should I just copy the new CPU values?~~

Update to match cuda values 👍🏼

File: src/transformers/models/rf_detr/image_processing_rf_detr.py (hunk 7 above)
Comment (guarin) [reply src/transformers/models/rf_detr/image_processing_rf_detr.py:557]: This if a fix from rf-detr: https://github.com/roboflow/rf-detr/blob/5547a9360c19df7ff8cedcd29f841f7de01f2f75/src/rfdetr/detr.py#L2480-L2481

File: src/transformers/models/conditional_detr/image_processing_conditional_detr.py (hunk 3 above)
Comment (guarin) [reply src/transformers/models/conditional_detr/image_processing_conditional_detr.py]: Aah dang, good catch. Will update including reverting the arg names

File: src/transformers/models/conditional_detr/image_processing_conditional_detr.py (hunk 4 above)
Comment (guarin) [reply src/transformers/models/conditional_detr/image_processing_conditional_detr.py:744]: Good idea! I also wonder if there is generally a better way to handle grouping and batched preprocessing, it is a bit messy right now.

File: src/transformers/models/rf_detr/modular_rf_detr.py (hunk 5 above)
Comment (guarin) [reply src/transformers/models/rf_detr/modular_rf_detr.py:205]: rfdetr does rescale then resize while detr does resize then rescale

File: src/transformers/image_transforms.py (hunk 6 above)
Comment (guarin) [reply src/transformers/image_transforms.py:1041]: because with `disable_grouping==False` we always returned a list instead of unsqueeze. The reason is that paired values are not necessarily tensors.

File: src/transformers/models/conditional_detr/image_processing_conditional_detr.py (hunk 4 above)
Comment (zucchini-nlp) [reply src/transformers/models/conditional_detr/image_processing_conditional_detr.py:744]: Totally agreed and it's not very easy for contribs imo when it comes to custom/nested processing. I don't know how we can do better, afaik `torch.nested_tensors` doesn't work yet for all ops. Prob Pablo's vision-kernels can help to build smth

File: src/transformers/models/rf_detr/image_processing_rf_detr.py (hunk 7 above)
Comment (zucchini-nlp) [reply src/transformers/models/rf_detr/image_processing_rf_detr.py:557]: oke, can we mark PR as 🚨

File: src/transformers/models/conditional_detr/image_processing_conditional_detr.py [hunk 8]
```diff
@@ -735,69 +748,74 @@ def _preprocess(
...
         if do_pad:
             # depends on all resized image shapes so we need another loop
             if pad_size is not None:
                 padded_size = (pad_size.height, pad_size.width)
             else:
-                padded_size = get_max_height_width(images)
-
-            padded_images = []
-            padded_annotations = []
-            for image, annotation in zip(images, annotations if annotations is not None else [None] * len(images)):
```
Comment (zucchini-nlp) [inline src/transformers/models/conditional_detr/image_processing_conditional_detr.py:781]: `do_pad` can go under the same loop since we're reusing the rescaled images without "reorder then group again" code in between

**Review by zucchini-nlp — APPROVED** [review]
LGTM, lets run slow tests for rf_detr before merging

File: src/transformers/models/conditional_detr/image_processing_conditional_detr.py (hunk 8 above)
Comment (guarin) [reply src/transformers/models/conditional_detr/image_processing_conditional_detr.py:781]: Sadly not possible because of `padded_size = get_max_height_width(images)` which requires the first loop to be complete.

**Comment by guarin** [comment]
run-slow: conditional_detr, deformable_detr, detr, grounding_dino, rf_detr, rt_detr, yolos

**Review by molbap — APPROVED** [review]
Went through it!  sounds good to me as well. Let's merge soon? maybe a quick core eye from @vasqu ?

File: src/transformers/models/detr/image_processing_detr.py [hunk 9]
```diff
@@ -656,13 +662,19 @@ def pad(
             padding = [0, 0, padding_right, padding_bottom]
             image = tvF.pad(image, padding, fill=fill)
             if annotation is not None:
-                annotation = self._update_annotation_for_padded_image(
-                    annotation, original_size, padded_size, padding, update_bboxes
-                )
+                is_single_annotation = isinstance(annotation, dict)
+                annotations = [annotation] if is_single_annotation else annotation
+                annotations = [
+                    self._update_annotation_for_padded_image(
```
Comment (vasqu) [inline src/transformers/models/detr/image_processing_detr.py:668]: honestly dont see why we are not just updating `_update_annotation_for_padded_image` instead because we kind of workaround around it here 🤔

File: src/transformers/models/detr/image_processing_detr.py [hunk 10]
```diff
@@ -733,69 +746,74 @@ def _preprocess(
...
+                orig_size = stacked_images.shape[-2:]
+                stacked_images = self.resize(stacked_images, size=size, resample=resample)
+                if stacked_annotations is not None:
+                    stacked_annotations = [
+                        self.resize_annotation(
+                            annotation,
+                            orig_size=orig_size,
+                            target_size=stacked_images.shape[-2:],
+                        )
+                        for annotation in stacked_annotations
```
Comment (vasqu) [inline src/transformers/models/detr/image_processing_detr.py:779]: same for all these feels like, just don't see why we make this properly within all annotation functions to handle single and multiple to keep these loops out of here. we only want a single call here tbh and let downstream handle batches

File: tests/models/rf_detr/test_modeling_rf_detr.py
Comment (vasqu) [inline tests/models/rf_detr/test_modeling_rf_detr.py:1]: Ok so grouping introduces slight numeric differences? (which imo is kind of expected but just want confirmation)

**Review by vasqu — APPROVED** [review]
Imo just one design question but if you feel like it's not really appropriate feel free to ignore

File: src/transformers/models/detr/image_processing_detr.py (hunk 9 above)
Comment (guarin) [reply src/transformers/models/detr/image_processing_detr.py:668]: Ah this is for BC. Happy to change if we accept those kinds of minor bc issues (change interface of `_update_annotation_for_padded_image`)

File: tests/models/rf_detr/test_modeling_rf_detr.py
Comment (guarin) [reply tests/models/rf_detr/test_modeling_rf_detr.py:1]: Didn't observe any differences due to grouping, the change here is from this fix: https://github.com/huggingface/transformers/pull/48066#discussion_r3841835371

Happy to move the fix to a separate pr if preferred

Comment (vasqu) [reply tests/models/rf_detr/test_modeling_rf_detr.py:1]: Ah no thats fine then was just wondering

File: src/transformers/models/detr/image_processing_detr.py (hunk 9 above)
Comment (vasqu) [reply src/transformers/models/detr/image_processing_detr.py:668]: I mean we could keep BC by accepting both no? It would be an additional function. Just don't think it makes much sense to make so many workarounds for this

Comment (guarin) [reply src/transformers/models/detr/image_processing_detr.py:668]: If I change `_update_annotation_for_padded_image` (or `self.resize_annotation`) to accept a list then subclasses will break as they don't expect a list input. The probability that someone subclassed these methods is imo low. 

Alternatively I can create an additional function as you mentioned. Then we would get more methods `_update_annotation_list_for_padded_image` etc. It is also fine just bloats the code a bit.

Comment (vasqu) [reply src/transformers/models/detr/image_processing_detr.py:668]: Nah then dw, we can keep as is for now. Sounds a bit like a headache so lets play it safe

**Comment by guarin** [comment]
run-slow: conditional_detr
